File: extension/core/src/main/java/io/holunda/camunda/bpm/data/reader/VariableMapReader.java
package io.holunda.camunda.bpm.data.reader;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.variable.VariableMap;

import java.util.Objects;
import java.util.Optional;

/**
 * Allows reading multiple variable values from {@link VariableMap#getValue(String, Class)}.
 */
public class VariableMapReader implements VariableReader {

  private final VariableMap variableMap;

  /**
   * Constructs the reader using the specified variable map.
   *
   * @param variableMap map to operate on.
   */
  public VariableMapReader(VariableMap variableMap) {
    this.variableMap = variableMap;
  }

  @Override
  public <T> Optional<T> getOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(variableMap).getOptional();
  }

  @Override
  public <T> T get(VariableFactory<T> variableFactory) {
    return variableFactory.from(variableMap).get();
  }

  @Override
  public <T> T getLocal(VariableFactory<T> variableFactory) {
    return variableFactory.from(variableMap).getLocal();
  }

  @Override
  public <T> Optional<T> getLocalOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(variableMap).getLocalOptional();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    VariableMapReader that = (VariableMapReader) o;

    return Objects.equals(variableMap, that.variableMap);
  }

  @Override
  public int hashCode() {
    return variableMap != null ? variableMap.hashCode() : 0;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/reader/VariableMapReader.kt
package io.holunda.camunda.bpm.data.reader

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.variable.VariableMap
import java.util.*

/**
 * Allows reading multiple variable values from [VariableMap.getValue].
 * @param variableMap map to operate on.
 */
class VariableMapReader(private val variableMap: VariableMap) : VariableReader {

  override fun <T> getOptional(variableFactory: VariableFactory<T>): Optional<T> {
    return variableFactory.from(variableMap).getOptional()
  }

  override operator fun <T> get(variableFactory: VariableFactory<T>): T {
    return variableFactory.from(variableMap).get()
  }

  override fun <T> getLocal(variableFactory: VariableFactory<T>): T {
    return variableFactory.from(variableMap).getLocal()
  }

  override fun <T> getLocalOptional(variableFactory: VariableFactory<T>): Optional<T> {
    return variableFactory.from(variableMap).getLocalOptional()
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as VariableMapReader
    return variableMap == that.variableMap
  }

  override fun hashCode(): Int {
    return variableMap.hashCode()
  }
}
